### PR TITLE
P2p leaks

### PR DESCRIPTION
**Pull request by y0sher**

### CODE REVIEW

**Comment by y0sher** [comment]
test will pass after we merge #466

File: p2p/connectionpool/connectionpool.go [hunk 1]
```diff
@@ -178,7 +178,9 @@ func (cp *ConnectionPool) GetConnection(address string, remotePub crypto.PublicK
 	// where it is possible that the connection will be established and all registered channels will be notified before
 	// the current registration
 	cp.pendMutex.Lock()
+	pendChan := make(chan dialResult)
```
Comment (gavraz) [inline p2p/connectionpool/connectionpool.go]: Move this line to be right before the line that uses the pendChan.

File: p2p/net/network.go [hunk 2]
```diff
@@ -302,7 +302,7 @@ func (n *Net) acceptTCP() {
 // SubscribeOnNewRemoteConnections returns new channel where events of new remote connections are reported
 func (n *Net) SubscribeOnNewRemoteConnections() chan NewConnectionEvent {
 	n.regMutex.Lock()
-	ch := make(chan NewConnectionEvent)
+	ch := make(chan NewConnectionEvent, n.config.SwarmConfig.RandomConnections)
```
Comment (gavraz) [inline p2p/net/network.go]: Do you really need this? if so, is it enough?

File: p2p/net/conn.go [hunk 3]
```diff
@@ -152,10 +152,10 @@ func (c *FormattedConnection) Send(m []byte) error {
 
 // Close closes the connection (implements io.Closer). It is go safe.
 func (c *FormattedConnection) Close() {
-	c.closeOnce.Do(func() {
-		atomic.AddInt32(&c.closed, 1)
-		c.closeChan <- struct{}{}
-	})
+	if atomic.LoadInt32(&c.closed) != 1 {
```
Comment (gavraz) [inline p2p/net/conn.go]: You can have a race here

File: p2p/net/conn.go [hunk 4]
```diff
@@ -165,8 +165,12 @@ func (c *FormattedConnection) Closed() bool {
 
 func (c *FormattedConnection) shutdown(err error) {
 	c.logger.Info("(%v) shutdown. id=%s err=%v", c.remotePub.String(), c.id, err)
+	if atomic.LoadInt32(&c.closed) != 1 {
```
Comment (gavraz) [inline p2p/net/conn.go]: You can have a race here

File: p2p/swarm.go [hunk 5]
```diff
@@ -345,28 +346,25 @@ func (s *swarm) RegisterProtocol(protocol string) chan service.Message {
 // Shutdown sends a shutdown signal to all running services of swarm and then runs an internal shutdown to cleanup.
 func (s *swarm) Shutdown() {
```
Comment (gavraz) [inline p2p/swarm.go:325]: Consider adding a comment that states that this function cannot be called twice (and also isn't thread-safe)

File: p2p/swarm.go [hunk 6]
```diff
@@ -403,7 +401,8 @@ Loop:
 		case con := <-closing:
 			go s.retryOrReplace(con.RemotePublicKey()) //todo notify dht?
 		case nce := <-newConnEvents:
-			go func(nce net.NewConnectionEvent) { s.dht.Update(nce.Node); s.addIncomingPeer(nce.Node.PublicKey()) }(nce)
```
Comment (gavraz) [inline p2p/swarm.go:374]: As we talked, think if you really need the "go" call inside. Generally, prefer buffering over go calls.

File: p2p/swarm.go [hunk 7]
```diff
@@ -403,7 +401,8 @@ Loop:
 		case con := <-closing:
 			go s.retryOrReplace(con.RemotePublicKey()) //todo notify dht?
```
Comment (gavraz) [inline p2p/swarm.go]: Rename to retryOrDisconnect

File: p2p/swarm.go [hunk 8]
```diff
@@ -543,9 +547,9 @@ func (s *swarm) onRemoteClientMessage(msg net.IncomingMessageEvent) error {
 // ProcessProtocolMessage passes an already decrypted message to a protocol.
 func (s *swarm) ProcessProtocolMessage(sender node.Node, protocol string, data service.Data) error {
 	// route authenticated message to the reigstered protocol
-	s.protocolHandlerMutex.RLock()
+	s.protocolHandlerMutex.Lock()
```
Comment (gavraz) [inline p2p/swarm.go]: Why? You are just reading. Writes cannot be made once the read lock is acquired.

**Review by gavraz — DISMISSED** [review]

File: p2p/swarm.go (hunk 6 above)
Comment (y0sher) [reply p2p/swarm.go:374]: both send on buffered chans so should be ok.

File: p2p/swarm.go (hunk 5 above)
Comment (y0sher) [reply p2p/swarm.go:325]: the contract is should be only called once. it doesn't matter from which thread.
so as long as it used once it is considered thread safe.

File: p2p/swarm.go (hunk 8 above)
Comment (y0sher) [reply p2p/swarm.go]: you're right

File: p2p/net/conn.go [hunk 9]
```diff
@@ -167,9 +166,15 @@ func (c *FormattedConnection) Closed() bool {
 }
 
 func (c *FormattedConnection) shutdown(err error) {
-	c.logger.Info("(%v) shutdown. id=%s err=%v", c.remotePub.String(), c.id, err)
+	atomic.StoreInt32(&c.closed, 1)
```
Comment (almogdepaz) [inline p2p/net/conn.go]: lose the atomic int and pass a nil err to differentiate , its cleaner

File: p2p/connectionpool/connectionpool.go [hunk 10]
```diff
@@ -53,14 +49,36 @@ func NewConnectionPool(network networker, lPub p2pcrypto.PublicKey) *ConnectionP
...
+func (cp *ConnectionPool) OnClosedConnection(c net.Connection) {
+	if cp.isShuttingDown() {
+		return
+	}
+	cp.handleClosedConnection(c)
+}
+
+func (cp *ConnectionPool) isShuttingDown() bool {
+	var isd bool
+	cp.connMutex.Lock()
```
Comment (antonlerner) [inline p2p/connectionpool/connectionpool.go]: can use rlock

File: p2p/net/conn.go [hunk 11]
```diff
@@ -167,9 +166,15 @@ func (c *FormattedConnection) Closed() bool {
...
+	select {
+		case <-c.closeChan:
+			// if we called close then we're in charge of reporting it.
+			break
+		default:
+			c.networker.publishClosingConnection(c)
```
Comment (antonlerner) [inline p2p/net/conn.go]: please refactor this

File: p2p/net/network.go [hunk 12]
```diff
@@ -154,18 +155,16 @@ func (n *Net) IncomingMessages() []chan IncomingMessageEvent {
...
 	n.clsMutex.Unlock()
-	return ch
 }
 
 func (n *Net) publishClosingConnection(connection Connection) {
 	n.clsMutex.RLock()
-	for _, c := range n.closingConnections {
-		c <- connection
+	for _, f := range n.closingConnections {
+		f(connection)
```
Comment (antonlerner) [inline p2p/net/network.go:167]: f can block? is this ok if it blocks? if not- maybe run f in go routine

File: p2p/swarm.go [hunk 13]
```diff
@@ -154,27 +155,43 @@ func newSwarm(ctx context.Context, config config.Config, newNode bool, persist b
 
 		protocolHandlers: make(map[string]chan service.Message),
 		network:          n,
-		cPool:            connectionpool.NewConnectionPool(n, l.PublicKey()),
 	}
 
 	s.dht = dht.New(l, config.SwarmConfig, s)
 
+	s.network.SubscribeOnNewRemoteConnections(s.onNewConnection)
```
Comment (almogdepaz) [inline p2p/swarm.go:162]: i would put all the handler registrations together would be easier for people not familiar with the code

File: p2p/net/network.go [hunk 14]
```diff
@@ -280,24 +279,17 @@ func (n *Net) acceptTCP() {
...
 func (n *Net) publishNewRemoteConnectionEvent(conn Connection, node node.Node) {
 	n.regMutex.RLock()
-	for _, c := range n.regNewRemoteConn {
-		select {
-		case c <- NewConnectionEvent{conn, node}:
-			continue
-		default:
-			// so we won't block on not listening chans
+	for _, f := range n.regNewRemoteConn {
+		 f(NewConnectionEvent{conn, node})
```
Comment (antonlerner) [inline p2p/net/network.go:291]: can f block? is this ok? if not - goroutine

**Review by antonlerner — APPROVED** [review]

**Review by almogdepaz — APPROVED** [review]

File: p2p/net/network.go (hunk 12 above)
Comment (y0sher) [reply p2p/net/network.go:167]: it should block.

**Pull request merged**